### DISCUSSION TITLE
Need help to find out drop in performance after updated to azure function durable extension v2.6.1 (from v2.5.1)

### DISCUSSION THREAD

**Discussion by siemenstan** [Azure/azure-functions-durable-extension#2358] [Q&A]
Hi,

I would like to ask for some help or guidance in checking why the drop in performance after updated my app (running inside azure kubernetes) with durable extension v2.6.1 (from v2.5.1), there is no any code change nor configuration changes apart from the durable extension library version change (Didn't update to the latest durable extension due to the issue filed in [discussion#2335](https://github.com/Azure/azure-functions-durable-extension/discussions/2335)).

Below is my app info:
Azure Region: West US
App info: C# .NET 6.0 on Azure Kubernetes (1 Orchestrator function with 3 activity chaining pattern)
ApplicationInsights Name: ncppfc01aiseus01
Storage Account Name: ncppfc01xenstrwus02                
> partitionCount=6, and manually scale to 6 pods to avoid rebalance during the loadtest. Here is the [host.json](https://github.com/Azure/azure-functions-durable-extension/files/10431794/host.json.txt)

Load test is done by using jmeter to fire 8000 requests per minute to the app, for 15 minutes. Result and (timing) response from each request is captured and benchmarked.
Load test session (with durable extension v2.6.1) is between **14-Jan-2023, 09:38:00 to 10:07:00 UTC**
  Sample orchestration instance id: `b921a973-8728-4c32-b37b-ac2b3513e721` ([AppInsights link](https://portal.azure.com/#blade/AppInsightsExtension/DetailsV2Blade/DataModel/%7B%22eventId%22:%22cc4b8074-93ef-11ed-a00c-c55ff87d7ddf%22,%22timestamp%22:%222023-01-14T09:42:35.4248106Z%22%7D/ComponentId/%7B%22Name%22:%22ncppfc01aiseus01%22,%22ResourceGroup%22:%22ncppfc01%22,%22SubscriptionId%22:%2230e1c5ff-d39c-428f-9bf8-1a52eff4246b%22%7D))

Load test session (with durable extension v2.5.1) is between **14-Jan-2023, 08:06:00 to 08:35:00 UTC**
  Sample orchestration instance id: `e9c62ccc-de90-4954-913a-e34645e40ec8` ([AppInsights link](https://portal.azure.com/#blade/AppInsightsExtension/DetailsV2Blade/DataModel/%7B%22eventId%22:%22f89d8cc7-93e2-11ed-b6f8-637e3aa7da96%22,%22timestamp%22:%222023-01-14T08:10:55.6867281Z%22%7D/ComponentId/%7B%22Name%22:%22ncppfc01aiseus01%22,%22ResourceGroup%22:%22ncppfc01%22,%22SubscriptionId%22:%2230e1c5ff-d39c-428f-9bf8-1a52eff4246b%22%7D))

Here is the AppInsights E2E timeline visualisation, I would like to understand why there orchestrator is slow when the app is updated with durable extension v2.6.1.
![image](https://user-images.githubusercontent.com/98590735/212820475-23858ead-47a3-4be0-9a90-29d7997b16a8.png)

I can see some high delay, PendingOrchestratorMessageLimitReached, and split brain traces in the trace message. The occurrence of such trace is much higher with durable extension v2.6.1 than v2.5.1 with same configuration and same loadtest condition. Is there any parameter we need to update after updated to durable extension v2.6.1?
![image](https://user-images.githubusercontent.com/98590735/212820769-1d9bbe6d-b492-49da-be08-676e8c22c3a6.png)  
![image](https://user-images.githubusercontent.com/98590735/212821023-b8b6664e-dd92-4f11-90aa-85e95153317a.png)
![image](https://user-images.githubusercontent.com/98590735/212821250-e6fa569a-1cf9-4873-80ad-e8ccf0c245e0.png)

Thank you

**Comment by cgillum** [comment]
It's hard to say what the source of the slowness, or whether it's even related to the specific version update. Both versions you're referring to are quite old, though, so I wonder if it would be better to try upgrading to v2.9.0 (the latest as of right now) instead of v2.6.1.

That said, split-brain errors are pretty much never expected, so something appears to be off. What configuration values do you have in your host.json file? Also, are you doing any auto-scaling of your environment? Split-brain was more common back when we used an older partition management scheme and when the number of instance hosting an app changed (e.g. scaling from 1 to 10), but we've since defaulted to a cooperative partition management strategy that should get rid of these types of problems. The other time I've seen split-brain is when multiple instances of the app are running on the same machine - i.e. the machine name environment variable is the same. This confuses our lease mechanism, though it's typically something you see when trying to run multiple instances on a local machine. I assume a containerized app wouldn't have this problem.

Lastly, is this problem persistent, or does it eventually go away? If it eventually goes away, that might suggest it is related to some change in the update, though I didn't see anything obvious when scanning the release notes for v2.6.0 and v2.6.1.

**Reply by siemenstan** [reply]
This is my app [host.json](https://github.com/Azure/azure-functions-durable-extension/files/10431794/host.json.txt). Only the few parameters are configured, the rest are pretty much all stick with default value and behavior.
```
{
"version": "2.0",
"extensions": {
  "durableTask": {
    "storageProvider": {
      "controlQueueBatchSize": 32,
      "controlQueueBufferThreshold": 256,
      "controlQueueVisibilityTimeout": "00:05:00",
      "maxQueuePollingInterval": "00:00:30",
      "partitionCount": 6,
      "workItemQueueVisibilityTimeout": "00:05:00",
      "connectionStringName": "AzureWebJobsStorage"
    },
    "maxConcurrentActivityFunctions": 30,
    "maxConcurrentOrchestratorFunctions": 30,
    "hubName": "%TaskHubName%"
  }
},
"logging": {
  "fileLoggingMode": "never",
  "console": {
    "isEnabled": "false"
  },
  "logLevel": {
    "default": "Debug",
    "Host.Results": "Debug",
    "Host.Aggregator": "Debug",
    "DurableTask.Core": "Debug"
  },
  "applicationInsights": {
    "samplingSettings": {
      "isEnabled": false,
      "maxTelemetryItemsPerSecond": 500
    }
  }
}
}
```

Auto-scaling is off during the load test, all the pods are manually scaled to tested loadtest configuration (which is from loadtest result with durable extension v2.5.1). We would like to upgrade the to latest durable extension but there is issue starting from v2.7.0 (till v2.9.0), reported in [issue#2336](https://github.com/Azure/azure-functions-durable-extension/issues/2336), so v2.6.1 is the latest version we can used now. But, for your information, I did the same loadtest with app updated to modified durable extension v2.9.0 (I modified with my hacky changes on top of the v2.9.0 code base), and the loadtest result is also no good. I also tried adjust some of the parameters value like controlQueueBufferThreshold, maxConcurrentActivityFunctions, and maxConcurrentOrchestratorFunctions but the result is still far from the performance we tested with durable extension v2.5.1, and I out of idea what else should I look into.

```
Load test session (with modified durable extension v2.9.0) is between 15-Jan-2023, 11:19:00 to 11:55:00 UTC
Sample orchestration instance id: 89f7935e-87b4-41b4-8fc5-76f058037053
```

Btw, back to the [app insights log](https://portal.azure.com#@a537e37f-5f7d-414a-90dc-66a7b74fbac4/blade/Microsoft_OperationsManagementSuite_Workspace/Logs.ReactView/resourceId/%2Fsubscriptions%2F30e1c5ff-d39c-428f-9bf8-1a52eff4246b%2FresourceGroups%2Fncppfc01%2Fproviders%2Fmicrosoft.insights%2Fcomponents%2Fncppfc01aiseus01/source/LogsBlade.AnalyticsShareLinkToQuery/q/H4sIAAAAAAAAA4WSTW7bMBCF9zoFoZUdlDIlJZDtJl3ViwBtErhFUHRjUNQ0JiKRLEn5J8hhcpacLENbigUUQbUiZ96bEb%252BZGjxxnltPrkjFPXjZwCjOWJZTltL0%252FCebzfPpnLGEMfY7Hn8mNTpAVR%252FpUzZnxUAfBb1U2EMJuA62uJxlKZ8VOZ0W2ZSeizyjZV6UlIuszC%252FSHIosjY9GB3YjBdzwBoLzZiuSX6CSxQ5E66VWyY%252BjoJevua16sdoKugNFoRfTULVVeCLS%252FWmfnvZX3rYQEfz4hsual7KWfr8E19befTokLPxtwfU32AkwoVZ3N%252FwB7iVsu6u3XEB3Fq3zullsQPXmCgyCAyUkuOiZbNdggQSACKcx5Es3CI5wT9HLADuaTHq9qHVbrZa6PkKRiowGkMbRv7rrDj5Zc3ci9C48O8QHE5pMtCVnry8Yfn2J77C9VA%252B3Vqzxjyz32n4H5%252FDd32Qj%252FRI4JqqYBM%252BhUuwMQiSl5VLF2AV2PqwLBBDdaLzGSlh0dGT0FR%252BrXICaLHrV%252BOTUBrAtpv%252Frvh0qBxUMYpUhfti%252Fj%252F3GarNa3Z3UoYa2FVhS7gcz4U5gwvNHICmu%252BRsfbLRaRAMAAA%253D%253D) with the durable extension v2.6.1, I am seeing the messages of this instance is moving to different control queue (control-03 to control-01 and back to control-03), is that expected? (I didn't see such situation with durable extension v2.5.1)
```
let start = datetime("2023-01-14T09:38:00.000Z"); let end = datetime("2023-01-14T10:07:00.000Z");
let instanceId = "b921a973-8728-4c32-b37b-ac2b3513e721";
let serviceName = "<removed>";
let shardName = "<removed>";
union isfuzzy=true
    availabilityResults,
    requests,
    exceptions,
    pageViews,
    traces,
    customEvents,
    dependencies
| where timestamp > start and timestamp < end
| where cloud_RoleInstance has shardName
| where * has instanceId
| extend eventName = tostring(customDimensions.EventName)
| extend operationName = tostring(customDimensions.OperationName)
| extend partitionId = tostring(customDimensions.prop__PartitionId)
| order by timestamp asc
| take 1000
```
![image](https://user-images.githubusercontent.com/98590735/213598264-937c994d-14da-4202-a3b1-741b219d400b.png)
![image](https://user-images.githubusercontent.com/98590735/213598587-e27a1fe3-cca4-43a1-a30a-485163ca9688.png)
![image](https://user-images.githubusercontent.com/98590735/213599088-4c1f7a30-be00-4afb-ba10-c56ce09ffe18.png)
